### PR TITLE
Add Support for Semi-sync cluster

### PR DESCRIPTION
**Pull request by heheh13**

### CODE REVIEW

File: pkg/controller/statefulset.go [hunk 1]
```diff
@@ -289,17 +290,56 @@ func getEnvsForMySQLCoordinatorContainer(db *api.MySQL) []core.EnvVar {
 		Name:  "GOVERNING_SERVICE_NAME",
 		Value: db.GoverningServiceName(),
 	})
-	if db.UsesGroupReplication() {
+	//if db.UsesGroupReplication() {
```
Comment (spectro30) [inline pkg/controller/statefulset.go:293]: remove commented code

File: pkg/controller/statefulset.go [hunk 2]
```diff
@@ -289,17 +290,56 @@ func getEnvsForMySQLCoordinatorContainer(db *api.MySQL) []core.EnvVar {
...
+				envList = append(envList, core.EnvVar{
+					Name:  "ErrantTransactionRecoveryPolicy",
+					Value: string(api.ErrantTransactionRecoveryPolicyClone),
+				})
+			}
+		}
+
+		// insert semisyc vars
+
+		if db.Spec.Topology.SemiSync != nil {
```
Comment (spectro30) [inline pkg/controller/statefulset.go:332]: add nil check for `SemiSyncMasterTimeout` and `SemiSyncMasterWaitFoSlaveCount`

**Review by spectro30 — CHANGES_REQUESTED** [review]

File: pkg/controller/health.go [hunk 3]
```diff
@@ -244,6 +265,35 @@ func closeClientEngine(engine *xorm.Engine) {
...
 
+func (c *Controller) checkSemiSyncClusterHealth(ctx context.Context, engine *xorm.Engine) (bool, error) {
+	session := engine.NewSession().Context(ctx)
+	defer func(session *xorm.Session) {
+		err := session.Close()
+		if err != nil {
+			klog.Error(err)
+		}
+	}(session)
+	// ready, critical , not ready..
```
[the remark this answers is not in the document]
Comment (heheh13) [reply pkg/controller/health.go:276]: random comments while coding.
removed

File: pkg/controller/health.go [hunk 4]
```diff
@@ -416,6 +465,24 @@ func (c *Controller) updateConditionsForNotHealthy(ctx context.Context, db *api.
...
 
+func (c *Controller) updateConditionsForCritical(ctx context.Context, db *api.MySQL, podName string, reason error) error {
+	err := c.updateMySQLStatusConditions(ctx, db,
+		kmapi.Condition{
+			Type:    api.DatabaseReady,
+			Status:  core.ConditionFalse,
+			Reason:  api.SomeReplicasAreNotReady,
+			Message: fmt.Sprintf("instance %s in not accepting connection %v", podName, reason),
+		},
+		kmapi.Condition{
```
[the remark this answers is not in the document]
Comment (heheh13) [reply pkg/controller/health.go]: udpated

File: pkg/controller/health.go [hunk 5]
```diff
@@ -109,6 +109,7 @@ func (c *Controller) CheckMySQLHealthOnce() {
 
 			// loop over all pods and check db health and update db conditions
 			replicaNotReadyCount := 0
+			semiSync := 0
 			for _, pod := range dbPods {
 
 				dns := HostDNS(db, pod.ObjectMeta)
```
[the remark this answers is not in the document]
Comment (heheh13) [reply pkg/controller/health.go:115]: I think it will be better to rework in the next release.

**Review by spectro30 — APPROVED** [review]
LGTM

**Pull request merged**